### PR TITLE
Add Outbox docs

### PR DESCRIPTION
**Pull request by yaron2**
Add documentation for the outbox feature.

Closes https://github.com/dapr/docs/issues/3733.

### CODE REVIEW

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
+---
+type: docs
+title: "How-To: Enable the transactional outbox pattern"
+linkTitle: "How-To: Enable the transactional outbox pattern"
+weight: 400
+description: "Commit a single transaction across a state store and pub/sub message broker"
+---
+
+The transactional outbox pattern is a well known powerful design pattern for sending notifications about a change in a system's state using a single transaction that spans across the database and message broker used to deliver the notification. Developers are faced with many difficult technical challanges when trying to implement this pattern on their own, which often involves writing error-prone central coordination managers that end up supporting a combination of one or two databases and message brokers at most.
```
Comment (msfussell) [inline daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]:
```suggestion
The transactional outbox pattern is a well known design pattern for sending notifications about a change in an application's state using a single transaction that spans across the database and the message broker used to deliver the notification. Developers are faced with many difficult technical challenges when trying to implement this pattern on their own, which often involves writing error-prone central coordination managers that end up supporting a combination of one or two databases and message brokers at most.

An example scenario for using the outbox pattern is writing a new user record to an account database and then sending a notification message that the account was successfully created. The user only wants to receive a message if this was successful.
```

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+With Dapr's outbox support, developers can notify subscribers when a system's state is created or updated when calling Dapr's [transactions API]({{< ref "state_api.md#state-transactions" >}}).
```
Comment (msfussell) [inline daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]:
```suggestion
With Dapr's outbox support, you can notify subscribers when an application's state is created, updated or deleted when calling Dapr's [transactions API]({{< ref "state_api.md#state-transactions" >}}).
```

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    value: "myOutboxPubsub"
+  - name: outboxDiscardWhenMissingState #Optional. Defaults to false
+    value: false
+```
+
+### Metadata fields
+
+| Name                | Required    | Default Value | Description                                            |
+| --------------------|-------------|---------------|------------------------------------------------------- |
+| outboxPublishPubsub | Yes         | N/A           | Sets the name of the pub/sub to use when publishing state changes
```
Comment (msfussell) [inline daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]:
```suggestion
| outboxPublishPubsub | Yes         | N/A           | Sets the name of the pub/sub component to  deliver the notifications when publishing state changes
```

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+| outboxPublishTopic  | Yes         | N/A           | Sets the topic to send the state changes to on the pub/sub configured with `outboxPublishPubsub`
```
Comment (msfussell) [inline daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]:
```suggestion
| outboxPublishTopic  | Yes         | N/A           | Sets the topic to send the state changes to on the pub/sub component configured with `outboxPublishPubsub`
```

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+| outboxPubsub        | No          | `outboxPublishPubsub`           | Sets the pub/sub to use for Dapr to coordinate the state and pub/sub transactions. If not set, the pub/sub configured with `outboxPublishPubsub` is used. This is useful if you want to separate the pub/sub used to send the state changes from the one used to coordinate the transaction
```
Comment (msfussell) [inline daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]:
```suggestion
| outboxPubsub        | No          | `outboxPublishPubsub`           | Sets the pub/sub component to use by Dapr to coordinate the state and pub/sub transactions. If not set, the pub/sub component configured with `outboxPublishPubsub` is used. This is useful if you want to separate the pub/sub component used to send the notification state changes from the one used to coordinate the transaction.
```

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+| outboxDiscardWhenMissingState  | No         | `false`           | By setting `outboxDiscardWhenMissingState` to `true`, Dapr will discard the transaction if it cannot find the state in the database and not retry again
```
Comment (msfussell) [inline daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]:
```suggestion
| outboxDiscardWhenMissingState  | No         | `false`           | By setting `outboxDiscardWhenMissingState` to `true`, Dapr will discard the transaction if it cannot find the state in the database and does not retry again.
```

Comment (msfussell) [reply daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]: When should this be used? This explains what is does, but also describing why you would set this is needed.

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+description: "Commit a single transaction across a state store and pub/sub message broker"
+---
+
+The transactional outbox pattern is a well known powerful design pattern for sending notifications about a change in a system's state using a single transaction that spans across the database and message broker used to deliver the notification. Developers are faced with many difficult technical challanges when trying to implement this pattern on their own, which often involves writing error-prone central coordination managers that end up supporting a combination of one or two databases and message brokers at most.
+
+With Dapr's outbox support, developers can notify subscribers when a system's state is created or updated when calling Dapr's [transactions API]({{< ref "state_api.md#state-transactions" >}}).
+
+## Requirements
+
+The outbox feature can be used with using any [transactional state store]({{< ref supported-state-stores >}}) supported by Dapr. All [Dapr pub/sub brokers]({{< ref supported-pubsub >}}) are supported with the outbox feature.
```
Comment (msfussell) [inline daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]:
```suggestion
The outbox feature can be used with using any [transactional state store]({{< ref supported-state-stores >}}) supported by Dapr. All [pub/sub brokers]({{< ref supported-pubsub >}}) are supported with the outbox feature.
```

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+*Note: Message brokers that work with the competing consumer pattern are encouraged to reduce the chances of duplicate events*
```
Comment (msfussell) [inline daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]:
```suggestion
{{% alert title="Note" color="primary" %}} 
Message brokers that work with the competing consumer pattern (for example [Apache Kafka]({{< ref setup-apache-kafka>}}) are encouraged to reduce the chances of duplicate events.
{{% /alert %}}
```

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md (hunk 7 above)
Comment (msfussell) [reply daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]: If these properties are set on a non-transactional state store, does the Dapr runtime throw an error on loading? If not can it to indicate these are are invalid?

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+| outboxPublishPubsub | Yes         | N/A           | Sets the name of the pub/sub to use when publishing state changes
+| outboxPublishTopic  | Yes         | N/A           | Sets the topic to send the state changes to on the pub/sub configured with `outboxPublishPubsub`
+| outboxPubsub        | No          | `outboxPublishPubsub`           | Sets the pub/sub to use for Dapr to coordinate the state and pub/sub transactions. If not set, the pub/sub configured with `outboxPublishPubsub` is used. This is useful if you want to separate the pub/sub used to send the state changes from the one used to coordinate the transaction
+| outboxDiscardWhenMissingState  | No         | `false`           | By setting `outboxDiscardWhenMissingState` to `true`, Dapr will discard the transaction if it cannot find the state in the database and not retry again
+
+### Combining outbox and non-outbox messages on the same state store
+
+If you want to use the same state store for sending both outbox and non-outbox messages, simply define two state store components, where one has the outbox feature and the other does not.
```
Comment (msfussell) [inline daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]:
```suggestion
If you want to use the same state store for sending both outbox and non-outbox messages, simply define two state store components that connect to the same state store, where one has the outbox feature and the other does not.
```

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md (hunk 7 above)
Comment (msfussell) [reply daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]: Is there any metadata setting that is **required** to be set on the pub/sub component? Conversely Is there anything that I should not set that may break the outbox?

Comment (msfussell) [reply daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]: Can the In-memory message broker be used?

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md (hunk 9 above)
Comment (msfussell) [reply daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]: Two questions
1) Is it better to use MySQL as an example here than Redis, given that Redis is not a "True" transactional state store
2) If I do use Redis, could I use the same Redis instance for both the state store and the message broker? Is so, then I would mention this.

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md (hunk 4 above)
Comment (msfussell) [reply daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]: Can I configure what is sent in the message body? If not, what is sent in the message body for the topic?

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md (hunk 2 above)
Comment (msfussell) [reply daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]: Add the text below....

*****
```
The diagram below is an overview of how the outbox pattern works:
1) Service A saves/updates state to the state store using a transaction.
2) A message is written to the broker under the same transaction. When the message is successfully delivered to the message broker the transaction completes ensuring the state and message are transacted together.
3) The message broker delivers the message topic to any subscribers, in this case Service B.

<img src="/images/state-management-outbox.png" width=800 alt="Diagram showing the steps of the outbox pattern">
```

****

Comment (msfussell) [reply daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]: I will send you a PNG file to include into this PR and check into the static/images folder

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+  metadata:
+  - name: redisHost
+    value: localhost:6379
+  - name: redisPassword
+    value: ""
+  - name: outboxPublishPubsub # Required
+    value: "mypubsub"
+  - name: outboxPublishTopic # Required
+    value: "newOrder"
+```
```
Comment (msfussell) [inline daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]: Add this at the end

```
## Demo

Watch [this video for an overview of the outbox pattern](https://youtu.be/rTovKpG0rhY?t=1338):

<div class="embed-responsive embed-responsive-16by9">
<iframe width="360" height="315" src="https://youtu.be/rTovKpG0rhY?t=1338" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
```

**Review by msfussell — COMMENTED** [review]
Here is my review and provided updates

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md (hunk 2 above)
Comment (yaron2) [reply daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]: The suggested change is incorrect, outbox is about notifying when data has been created or inserted, not deleted.

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md (hunk 4 above)
Comment (yaron2) [reply daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]: No, you can't and shouldn't be able to configure what's in the message body. The message body is what is saved to the state store. I added language to make that clear, even though it's likely an overkill.

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md (hunk 10 above)
Comment (yaron2) [reply daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]: Done

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md (hunk 9 above)
Comment (yaron2) [reply daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]: Yes, I switched to MySQL as its a better example.

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md (hunk 7 above)
Comment (yaron2) [reply daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]: > If these properties are set on a non-transactional state store, does the Dapr runtime throw an error on loading? If not can it to indicate these are are invalid?

The user will get an error when trying to use the transactions API on a non-transactional state store.

Comment (yaron2) [reply daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]: > Is there any metadata setting that is required to be set on the pub/sub component? Conversely Is there anything that I should not set that may break the outbox?

No, nothing is required on the pub/sub component and there's no way to break it.

Comment (yaron2) [reply daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]: > Can the In-memory message broker be used?

Yes

**Comment by yaron2** [comment]
@msfussell all feedback has been addressed

**Comment by msfussell** [comment]
@hhunter-ms - LGTM. Any feedback on this?

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md [hunk 11]
```diff
@@ -0,0 +1,108 @@
+---
+type: docs
+title: "How-To: Enable the transactional outbox pattern"
+linkTitle: "How-To: Enable the transactional outbox pattern"
+weight: 400
+description: "Commit a single transaction across a state store and pub/sub message broker"
+---
+
+The transactional outbox pattern is a well known design pattern for sending notifications about a change in an application's state using a single transaction that spans across the database and the message broker used to deliver the notification. Developers are faced with many difficult technical challenges when trying to implement this pattern on their own, which often involves writing error-prone central coordination managers that end up supporting a combination of one or two databases and message brokers at most.
```
Comment (hhunter-ms) [inline daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]:
```suggestion
The transactional outbox pattern is a well known design pattern for sending notifications regarding changes in an application's state. The transactional outbox pattern uses a single transaction that spans across the database and the message broker delivering the notification. 

Developers are faced with many difficult technical challenges when trying to implement this pattern on their own, which often involves writing error-prone central coordination managers that, at most, support a combination of one or two databases and message brokers.
```

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+An example scenario for using the outbox pattern is writing a new user record to an account database and then sending a notification message that the account was successfully created. The user only wants to receive a message if this was successful.
```
Comment (hhunter-ms) [inline daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]:
```suggestion
For example, you can use the outbox pattern to:  
1. Write a new user record to an account database. 
1. Send a notification message that the account was successfully created. 
```

I feel like the example scenario might help the doc flow better if moved under line 13...

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+
+With Dapr's outbox support, you can notify subscribers when an application's state is created or updated when calling Dapr's [transactions API]({{< ref "state_api.md#state-transactions" >}}).
+
+The diagram below is an overview of how the outbox feature works:
+
+1) Service A saves/updates state to the state store using a transaction.
+2) A message is written to the broker under the same transaction. When the message is successfully delivered to the message broker the transaction completes ensuring the state and message are transacted together.
```
Comment (hhunter-ms) [inline daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]:
```suggestion
2) A message is written to the broker under the same transaction. When the message is successfully delivered to the message broker, the transaction completes, ensuring the state and message are transacted together.
```

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+3) The message broker delivers the message topic to any subscribers, in this case Service B.
```
Comment (hhunter-ms) [inline daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]:
```suggestion
3) The message broker delivers the message topic to any subscribers - in this case, Service B.
```

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+
+<img src="/images/state-management-outbox.png" width=800 alt="Diagram showing the steps of the outbox pattern">
+
+## Requirements
+
+The outbox feature can be used with using any [transactional state store]({{< ref supported-state-stores >}}) supported by Dapr. All [pub/sub brokers]({{< ref supported-pubsub >}}) are supported with the outbox feature.
+
+{{% alert title="Note" color="primary" %}} 
+Message brokers that work with the competing consumer pattern (for example [Apache Kafka]({{< ref setup-apache-kafka>}}) are encouraged to reduce the chances of duplicate events.
```
Comment (hhunter-ms) [inline daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]:
```suggestion
Message brokers that work with the competing consumer pattern (for example, [Apache Kafka]({{< ref setup-apache-kafka>}}) are encouraged to reduce the chances of duplicate events.
```

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+  - name: outboxDiscardWhenMissingState #Optional. Defaults to false
+    value: false
+```
+
+### Metadata fields
+
+| Name                | Required    | Default Value | Description                                            |
+| --------------------|-------------|---------------|------------------------------------------------------- |
+| outboxPublishPubsub | Yes         | N/A           | Sets the name of the pub/sub component to deliver the notifications when publishing state changes
+| outboxPublishTopic  | Yes         | N/A           | Sets the topic to send the state changes to on the pub/sub configured with `outboxPublishPubsub`. The message body will be a state transaction item for an insert or update operation
```
Comment (hhunter-ms) [inline daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]:
```suggestion
| outboxPublishTopic  | Yes         | N/A           | Sets the topic that receives the state changes on the pub/sub configured with `outboxPublishPubsub`. The message body will be a state transaction item for an `insert` or `update` operation
```

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+| outboxPubsub        | No          | `outboxPublishPubsub`           | Sets the pub/sub component to use by Dapr to coordinate the state and pub/sub transactions. If not set, the pub/sub component configured with `outboxPublishPubsub` is used. This is useful if you want to separate the pub/sub component used to send the notification state changes from the one used to coordinate the transaction
```
Comment (hhunter-ms) [inline daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]:
```suggestion
| outboxPubsub        | No          | `outboxPublishPubsub`           | Sets the pub/sub component used by Dapr to coordinate the state and pub/sub transactions. If not set, the pub/sub component configured with `outboxPublishPubsub` is used. This is useful if you want to separate the pub/sub component used to send the notification state changes from the one used to coordinate the transaction
```

File: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+| outboxDiscardWhenMissingState  | No         | `false`           | By setting `outboxDiscardWhenMissingState` to `true`, Dapr will discard the transaction if it cannot find the state in the database and does not retry again. This setting can be useful if the state store data has been deleted for any reason before Dapr was able to deliver the message and you would like Dapr to drop the items from the pub/sub and stop retrying to fetch the state
```
Comment (hhunter-ms) [inline daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md]:
```suggestion
| outboxDiscardWhenMissingState  | No         | `false`           | By setting `outboxDiscardWhenMissingState` to `true`, Dapr discards the transaction if it cannot find the state in the database and does not retry. This setting can be useful if the state store data has been deleted for any reason before Dapr was able to deliver the message and you would like Dapr to drop the items from the pub/sub and stop retrying to fetch the state
```

**Review by hhunter-ms — DISMISSED** [review]
quick grammar review

**Comment by yaron2** [comment]
> quick grammar review

@hhunter-ms changes made.

**Review by msfussell — APPROVED** [review]
LGTM

**Pull request merged**